### PR TITLE
[fortinet_fortigate] Properly handle firewall URLs

### PR DESCRIPTION
**Pull request by dwhyrock**
## Proposed commit message
This fixes a bug where URLs were not parsed correctly.  It adds in the `uri_parts` processor, as well as adding in a painless script that recursively removes empty fields/values.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [ ] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

- [ ] 

## Related issues

- Relates https://github.com/elastic/sdh-beats/issues/4777

### CODE REVIEW

**Comment by dwhyrock** [comment]
Hey @efd6 , thanks so much for the [diff suggestion in the SDH](https://github.com/elastic/sdh-beats/issues/4777#issuecomment-2141792880)!

@taylor-swanson @pkoutsovasilis I may need some assistance on this one.  (As it's late on Friday, I wanted to get this down while it's still fresh in my head).

I added Dan's changes in my branch, and the URL parsing looks better for the most part, but some other tests are failing with some odd results.

Some troublesome ones where it is removing the `domain` include the following:

```
             "url": {
-                "domain": "www.dailymotion.com",
+                "original": "/",
                 "path": "/"
             }
```
and
```
             "url": {
-                "domain": "172.16.200.55",
+                "extension": "pdf",
+                "original": "/ips/sig1.pdf",
                 "path": "/ips/sig1.pdf"
             }
```

Another odd one that occurs in a couple places is where it inserts a `"url"` entry in the `fortinet.firewall` block, but it only includes the value `"/"`.
```
            "fortinet": {
                "firewall": {
                    "action": "block",
                    "appid": "16072",
                    "apprisk": "elevated",
                    "dstintfrole": "wan",
                    "incidentserialno": "1962906681",
                    "sessionid": "4414",
                    "srcintfrole": "lan",
                    "subtype": "app-ctrl",
                    "type": "utm",
 +                  "url": "/",
                    "vd": "root"
                }
            },
```

There are a lot of good results, like the one for the example log from the SDH looks good:
```
+            "url": {
+                "domain": "clientservices.googleapis.com",
+                "original": "http://clientservices.googleapis.com/chrome-variations/seed?osname=win&channel=stable&milestone=124",
+                "path": "/chrome-variations/seed",
+                "query": "osname=win&channel=stable&milestone=124",
+                "scheme": "http"
+            }
```

In a couple places, it seems like it will "clean up" some empty `_tmp` values that are likely not needed(?)
```
-            "_tmp": {
-                "user": {}
-            },
             "destination": {
                 "ip": "10.10.10.10"
             },
```

I've been playing around with the painless script a bit, but mostly causing more problems so far, lol.  (First time using painless).

Curious to hear if anyone has any thoughts on this one.  Thanks!

I've included the `elastic-package test pipeline` test output w/ the changes in this PR below:
[test-output.txt](https://github.com/user-attachments/files/15518535/test-output.txt)

**Comment by efd6** [comment]
Numbered by examples above.

1. The change in the `url` field in the first example is because there is a rename processor in the utm pipeline that is setting `url.domain` from `fortinet.firewall.host` before the `uri_parts` processor. This is thus being clobbered. This will need some domain-sensible logic to decide what to do. The simplest (though not necessarily correct approach would be to concat the `fortinet.firewall.host` and `fortinet.firewall.url` values before the `uri_parts` and then operate on that, but I don't know if the fields will always have these semantics).
2. Same issue here. This one supports the possibility of the approach above. Though looking at others, it seems there is neither rhyme nor reason to what is going on there.
3. The remaining `url` field in the `fortinet.firewall` group is due to the fact that we're no longer renaming. So the `fortinet.firewall.url` field should be removed after it is used.
4. -
5. That is intentional, otherwise a bunch of other things from the `url_parts` processor remain. But it should be done anyway.

**Comment by dwhyrock** [comment]
I've tried adding this to `utm.yml`:

```
    - uri_parts:
       field: fortinet.firewall.url
       target_field: url
       ignore_missing: true
+  - rename:
+      field: fortinet.firewall.hostname
+      target_field: url.domain
+      ignore_missing: true
+      override: true
+  - remove:
+      field: fortinet.firewall.url
+      ignore_missing: true
```

That way it overwrites the `url.domain` after the `uri_parts` processor runs.  It also removes the extra `fortinet.firewall.url` field.  It fixes some of the other issues, like:

this:
```
             "url": {
                 "domain": "www.dailymotion.com",
+                "original": "/",
                 "path": "/"
             }
```
instead of this:
```
             "url": {
-                "domain": "www.dailymotion.com",
+                "original": "/",
                 "path": "/"
             }
```

However, it seems to still add an `original` value there that's not exactly what was used to produce the fields in `url`, like so:

```
             "url": {
                 "domain": "www.dailymotion.com",
+                "original": "/",
                 "path": "/"
             }
```
and
```
             "url": {
                 "domain": "172.16.200.55",
+                "extension": "pdf",
+                "original": "/ips/sig1.pdf",
                 "path": "/ips/sig1.pdf"
             }
```

Is that acceptable?  It seems a bit odd to me if we were to try and overwrite any `.original` value.

**Comment by pkoutsovasilis** [comment]
yep the original part seems weird to me as well, maybe in the `uri_parts` processors you could set the field `keep_original` as `false` ([ref](https://www.elastic.co/guide/en/elasticsearch/reference/current/uri-parts-processor.html))

File: packages/fortinet_fortigate/data_stream/log/_dev/test/pipeline/test-fortinet.log [hunk 1]
```diff
@@ -53,3 +53,4 @@ date= desc="Object update request from device of FortiClient received" msg="Send
...
+<189>date=2024-05-23 time=15:20:07 devname="SMBFT......." devid="FGVxxxxxxxxx" eventtime=1716470406883543012 tz="+0200" logid="0000000013" type="traffic" subtype="forward" level="notice" vd="root" srcip=10.101.102.103 srcname="skoda.host.name" srcport=62430 srcintf="port2" srcintfrole="lan" dstip=10.250.186.99 dstname="clientservices.googleapis.com" dstport=80 dstintf="port3" dstintfrole="wan" srccountry="Reserved" dstcountry="United States" sessionid=1588040660 proto=6 action="deny" policyid=3 policytype="proxy-policy" poluuid="a783fde6-3a69-51eb-2d9e-4b559e26de3f" comment="PXIAA000" service="PRX-ALL" trandisp="noop" url="http://clientservices.googleapis.com/chrome-variations/seed?osname=win&channel=stable&milestone=124" agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36" duration=0 sentbyte=0 rcvdbyte=0 sentpkt=0 rcvdpkt=0 appcat="unscanned" crscore=30 craction=131072 crlevel="high" msg="Traffic denied because of explicit proxy policy" clustername="PROXY-xxx"
```
Comment (dwhyrock) [inline packages/fortinet_fortigate/data_stream/log/_dev/test/pipeline/test-fortinet.log:56]: FWIW, I updated the two IP addresses in this example log to be in the `10.0.0.0/8` subnet due to the `--generate` log message saying only certain IPs are allowed for Geo-location, which I took to mean everything outside of those non-RFC-1918 should be changed to RFC-1918 IPs.

**Comment by dwhyrock** [comment]
I'm now seeing errors like the following:

```
fortinet_fortigate/log test-fortinet-6-2.log:
[0] field "url.extension" is undefined
[1] field "url.scheme" is undefined
fortinet_fortigate/log test-fortinet-7-4.log:
[0] field "url.extension" is undefined
[1] field "url.query" is undefined
[2] field "url.scheme" is undefined
fortinet_fortigate/log test-fortinet.log:
[0] field "fortinet.firewall.clustername" is undefined
[1] field "url.query" is undefined
[2] field "url.scheme" is undefined
```

Should I put in processors to remove those fields sometime after the `uri_parts` processors are done?  It seems like the `uri_parts` processor is almost doing too good of a job breaking down the information, and that the `url` field doesn't expect it here.

**Comment by pkoutsovasilis** [comment]
> I'm now seeing errors like the following:
> 
> ```
> fortinet_fortigate/log test-fortinet-6-2.log:
> [0] field "url.extension" is undefined
> [1] field "url.scheme" is undefined
> fortinet_fortigate/log test-fortinet-7-4.log:
> [0] field "url.extension" is undefined
> [1] field "url.query" is undefined
> [2] field "url.scheme" is undefined
> fortinet_fortigate/log test-fortinet.log:
> [0] field "fortinet.firewall.clustername" is undefined
> [1] field "url.query" is undefined
> [2] field "url.scheme" is undefined
> ```
> 
> Should I put in processors to remove those fields sometime after the `uri_parts` processors are done? It seems like the `uri_parts` processor is almost doing too good of a job breaking down the information, and that the `url` field doesn't expect it here.

ok that's normal, now you have to add the new `fields` introduced by uri_parts processor inside `data_stream/log/fields/ecs.yml` and a new `clustername` field in `data_stream/log/fields/fields.yml` (as part of fortinet (group) -> firewall (group))

**Review by pkoutsovasilis — APPROVED** [review]
lgtm!

**Pull request merged**